### PR TITLE
feat: add required files check to test

### PR DESCRIPTION
**Pull request by alisd23**
Checks that all files in `required` folder exist in the current working directory and the contents are the same. Fails test if there is a missing or incorrect file.

### CODE REVIEW

File: scripts/test.js [hunk 1]
```diff
@@ -11,7 +13,70 @@ if (linterReport.warningCount > 0 || linterReport.errorCount > 0) {
...
 
-require('./bundle')
+// Check required files exist and content is correct
+function getFiles(cwd, globString) {
+    return new Promise((resolve, reject) => {
+        var globOptions = { cwd, dot: true };
+
+        glob(globString, globOptions, function(err, files) {
+            if (err) {
+                throw new Error('Required files check failed');
```
Comment (chrisprice) [inline scripts/test.js]: Should this just reject the promise?

File: scripts/test.js [hunk 2]
```diff
@@ -11,7 +13,70 @@ if (linterReport.warningCount > 0 || linterReport.errorCount > 0) {
     throw new Error('ESLint errors');
 }
 
-require('./bundle')
+// Check required files exist and content is correct
+function getFiles(cwd, globString) {
+    return new Promise((resolve, reject) => {
+        var globOptions = { cwd, dot: true };
+
+        glob(globString, globOptions, function(err, files) {
```
Comment (chrisprice) [inline scripts/test.js]: Use fat-arrows everywhere?

File: scripts/test.js [hunk 3]
```diff
@@ -11,7 +13,70 @@ if (linterReport.warningCount > 0 || linterReport.errorCount > 0) {
...
+                throw new Error('Required files check failed');
+            }
+            const fileDataPromise = Promise.all(
+                files.map(pathname => fs
+                  .readFile(path.join(cwd, pathname))
+                  .then(contents => ({ contents, filename: path.basename(pathname) }))
+                ))
+                .then(fileData => {
+                    const map = {};
+                    for (let file of fileData) {
```
Comment (chrisprice) [inline scripts/test.js]: const

File: scripts/test.js [hunk 4]
```diff
@@ -11,7 +13,70 @@ if (linterReport.warningCount > 0 || linterReport.errorCount > 0) {
...
+                    return map;
+                });
+
+            resolve(fileDataPromise);
+        });
+    });
+}
+
+function checkRequiredFiles() {
+    return getFiles(path.resolve(__dirname, '..'), 'required/*')
```
Comment (chrisprice) [inline scripts/test.js]: Just move the `..` into the glob?

File: scripts/test.js [hunk 5]
```diff
@@ -11,7 +13,70 @@ if (linterReport.warningCount > 0 || linterReport.errorCount > 0) {
...
+      .then(required => {
+          const fileGlob = Object.keys(required).join('|');
+          return getFiles(process.cwd(), `@(${fileGlob})`)
+            .then(candidates => ({
+                candidates,
+                required
+            }))
+            .catch(err => console.log('WEIRD ERROR', err));
```
Comment (chrisprice) [inline scripts/test.js]: Don't catch here

File: scripts/test.js [hunk 6]
```diff
@@ -11,7 +13,70 @@ if (linterReport.warningCount > 0 || linterReport.errorCount > 0) {
...
+
+            resolve(fileDataPromise);
+        });
+    });
+}
+
+function checkRequiredFiles() {
+    return getFiles(path.resolve(__dirname, '..'), 'required/*')
+      .then(required => {
+          const fileGlob = Object.keys(required).join('|');
```
Comment (chrisprice) [inline scripts/test.js]: Can this not be simplified quite a bit by just iterating over requried, using fs.readFile and blowing up as soon as it goes wrong?

**Comment by alisd23** [comment]
Looks better now

File: package.json [hunk 7]
```diff
@@ -27,6 +27,8 @@
     "eslint-config-standard": "^5.1.0",
     "eslint-plugin-promise": "^2.0.0",
     "eslint-plugin-standard": "^2.0.0",
+    "fs-promise": "^0.5.0",
```
Comment (chrisprice) [inline package.json]: Are you still using this?

File: scripts/test.js [hunk 8]
```diff
@@ -1,16 +1,63 @@
...
+
+// ====
+// REQUIRED FILES CHECK
+// ====
+const globOptions = {
+    cwd: required(),
+    dot: true
+};
+
+glob('*', globOptions, (err, files) => {
```
Comment (chrisprice) [inline scripts/test.js]: Do we not need to synchronise this with the other tests?

Comment (chrisprice) [inline scripts/test.js]: Maybe just make use of sync? https://www.npmjs.com/package/glob#globsyncpattern-options

File: scripts/test.js [hunk 9]
```diff
@@ -1,16 +1,63 @@
...
+    if (err) {
+        throw new Error('Required files check failed');
```
Comment (chrisprice) [inline scripts/test.js]: Don't swallow the error, it makes debugging a PITA. Maybe just concatenate it onto the error message?

File: scripts/test.js [hunk 10]
```diff
@@ -1,16 +1,63 @@
...
+    }
+    let failed = false;
+    files.forEach(filename => {
```
Comment (chrisprice) [inline scripts/test.js]: Use every to accumulate the result rather than a variable?

**Pull request merged**